src/libsyntax/parse/parser.rs

```diff
@@ -1344,9 +1344,31 @@ impl<'a> Parser<'a> {
                 }
             }
             self.expect(&token::CloseDelim(token::Paren))?;
+            let span = mk_sp(lo, self.prev_span.hi);
 
             if ts.len() == 1 && !last_comma {
-                TyKind::Paren(ts.into_iter().nth(0).unwrap())
+                let ty = ts.into_iter().nth(0).unwrap().unwrap();
+                match ty.node {
+                    // Accept `(Trait1) + Trait2 + 'a` for now,
+                    // but issue a future compatibility warning.
```

Comment:
I repurposed https://github.com/rust-lang/rust/issues/39318 as a tracking issue, I'll link to it.